src/__tests__/transform.test.ts

```diff
@@ -0,0 +1,272 @@
+  }
+
+  const result = container.resolve(Foo);
+
+  expect(result.value).toEqual("Transformed from bar");
+});
+
+test("Injecting with transform should work passing a parameter from the decorator", () => {
+  class Bar {
+    public repeat(str: string): string {
+      return str + str;
+    }
+  }
+
+  class BarTransform implements Transform<Bar, string> {
+    public transform(bar: Bar, str: string): string {
+      return bar.repeat(str);
+    }
+  }
+
+  @injectable()
+  class Foo {
+    constructor(
+      @injectWithTransform(Bar, BarTransform, "b") public value: string
+    ) {}
+  }
+
+  const result = container.resolve(Foo);
+
+  expect(result.value).toEqual("bb");
+});
+
+test("Injecting with transform should work passing parameters from the decorator", () => {
+  class Bar {
+    public concat(str1: string, str2: string): string {
+      return str1 + str2;
+    }
+  }
+
+  class BarTransform implements Transform<Bar, string> {
+    public transform(bar: Bar, str1: string, str2: string): string {
+      return bar.concat(str1, str2);
+    }
+  }
+
+  @injectable()
+  class Foo {
+    constructor(
+      @injectWithTransform(Bar, BarTransform, "a", "b") public value: string
+    ) {}
+  }
+
+  const result = container.resolve(Foo);
+
+  expect(result.value).toEqual("ab");
+});
+
+test("Transformation works with @autoInjectable", () => {
+  class Bar {}
+
+  class BarTransform {
+    public transform(): string {
+      return "Transformed from bar";
+    }
+  }
+
+  @autoInjectable()
+  class Foo {
+    constructor(
+      @injectWithTransform(Bar, BarTransform) public value?: string
+    ) {}
+  }
+
+  const myFoo = new Foo();
+
+  expect(myFoo.value).toBe("Transformed from bar");
+});
+
+test("Injecting all with transform should work", () => {
+  class Bar {}
+
+  class BarTransform {
+    public transform(): string {
+      return "Transformed from bar";
+    }
+  }
+
+  @injectable()
+  class Foo {
+    constructor(
+      @injectAllWithTransform(Bar, BarTransform) public value: string
+    ) {}
+  }
+
+  const result = container.resolve(Foo);
+
+  expect(result.value).toEqual("Transformed from bar");
+});
+
+test("Injecting all with transform should allow the transformer to act over an array", () => {
+  interface FooInterface {
+    bar: string;
+  }
+
+  class FooOne implements FooInterface {
+    public bar = "foo1";
+  }
+
+  class FooTwo implements FooInterface {
+    public bar = "foo2";
+  }
+
+  container.register<FooInterface>("FooInterface", {useClass: FooOne});
+  container.register<FooInterface>("FooInterface", {useClass: FooTwo});
+
+  class FooTransform implements Transform<FooInterface[], string> {
+    public transform(foos: FooInterface[]): string {
+      return foos.map(f => f.bar).reduce((acc, f) => acc + f);
+    }
+  }
+
+  @injectable()
+  class Bar {
+    constructor(
+      @injectAllWithTransform("FooInterface", FooTransform, "!!")
```

Comment:
Is the `!!` needed?

Reply:
It's just a string I inject after each string passed into the transformer. It could have been anything.